feat : 도메인 코드 생성

gutenLEE opened the pull request:
1. 카드덱
2. 카드 - 알파벳 카드, 숫자 카드
3. 유저 - 게이머, 딜러
4. 배팅 - 돈

- 블랙잭 게임 유저는 게임에 참여하려면 배팅할 돈과 카드가 필요하므로 합성 관계로 구성했습니다.

- AlphabetCard 나 NumberCard 클래스의 필드는 초기화 되면 값이 변경될 일이 없으므로 final 로 선언해야 됩니다. 
 하지만 52개의 카드 객체 생성은 런타임에 메서드로 해야된다면 final 을 선언할 수가 없는데
private 접근 제어로만 충분할까요?

- CardList.class 는 List<Card> 직접 참조보다 일급콜렉션 객체를 참조하여 List<Card> 조회나 검색 로직이 CardDeck 클래스에 섞이지 않도록 했습니다.

Commit:
feat : 도메인 코드 생성

1. 카드덱
2. 카드 - 알파벳 카드, 숫자 카드
3. 유저 - 게이머, 딜러
4. 배팅 - 돈

- 블랙잭 게임 유저는 게임에 참여하려면 배팅할 돈과 카드가 필요하므로 합성 관계로 구성했습니다.

- AlphabetCard 나 NumberCard 클래스의 필드는 초기화 되면 값이 변경될 일이 없으므로 final 로 선언해야 됩니다.
 하지만 52개의 카드 객체 생성은 런타임에 메서드로 해야된다면 final 을 선언할 수가 없는데
private 접근 제어로만 충분할까요?

- CardList.class 는 List<Card> 직접 참조보다 일급콜렉션 객체를 참조하여 List<Card> 조회나 검색 로직이 CardDeck 클래스에 섞이지 않도록 했습니다.

f-lab-maverick commented on src/card/Alphabet.java:
```diff
@@ -0,0 +1,8 @@
+package card;
+
+public enum Alphabet {
```
명확하게 enum으로 범위를 제한한 부분 좋습니다! Spring이나 char를 사용하지 않기만 해도 오타나 잘못된 값의 범위를 검증할 필요가 없어지는 장점이 있습니다

f-lab-maverick commented on src/card/CardDeck.java:
```diff
@@ -0,0 +1,15 @@
+
+    public CardDeck(CardList cardList) {
+        this.cardList = cardList;
+    }
```
CartDeck 자체가 카드를 담는 컬랙션으로 작동할 수 있지 않을까요? CardDeck과 CardList의 차이는 어떤것인가요?

gutenLEE replied:
카드덱의 역할은 카드 묶음 1) 넣고 2) 섞고 3) 나눠주기 에 집중해야 한다 생각했습니다!

카드덱의 역할을 정의하고 나니 
1. 카드 리스트 객체 초기화를 카드덱에서 하는게 맞을까? 
2. 카드덱 역할과 카드 리스트 쿼리 기능이 카드덱 클래스에 같이 존재하는게 맞을까? 
3. 카드덱의 카드는 언제든 교체될 수 있다는 것

라는 생각에 일급 컬렉션을 사용해봤습니다.

과도한 분리가 될수도 있을까요?

f-lab-maverick replied:
역할이 다르다면 지금 설계대로 진행하셔도 됩니다~ CardList 일급 컬랙션을 User클래스에 적용해보는것도 좋을것으로 보이네요.

이렇게 구현하시다가 `이렇게 구현하는게 맞을까?`라는 고민이 앞으로도 많이 되실거에요. 코드의 구현은 오답은 있지만 정답은 없습니다. 여러가지 구현 방법을 두고 장/단점에 대해 고민해보신 후 적용하는게 고민을 줄일 수 있는 길입니다. 추가적으로, 요구사항이 변경되며 변화가 필요할 때에는 리팩토링을 유연하게 할 수 있는 코드라면 더더욱 좋습니다.

f-lab-maverick commented on src/card/CardDeck.java:
```diff
@@ -0,0 +1,15 @@
+
+    public CardDeck(CardList cardList) {
+        this.cardList = cardList;
+    }
```
추가적으로, CardDeck을 생성할 때 생성자로 CardList를 받아야할 필요성이 있을까요?

f-lab-maverick commented on src/card/NumberCard.java:
```diff
@@ -0,0 +1,10 @@
+ * @since 2023/06/18
+ */
+public class NumberCard extends Card {
+    private Integer number;
```
1. 래퍼타입을 사용하신 이유가 있으실까요?
2. 카드의 숫자또한 범위값이 있지 않을까요? 모든 범위를 표현할 수 있는 Integer로 상태를 표현하게 된다면, 이 숫자를 사용할 때 마다 유효성에 대한 검증일 해야할 것입니다. 이런 경우 생성자에 검증 로직을 넣거나, 이런 원시 값을 래핑하여 사용할 수 있습니다. ex) `CardNumber` 클래스로 int값 래핑

f-lab-maverick commented on src/user/User.java:
```diff
@@ -0,0 +1,21 @@
+ * @author gutenlee
+ * @since 2023/06/18
+ */
+public abstract class User {
```
👍 딜러와 게이머의 공통점을 잘 잡았네요~

f-lab-maverick reviewed (changes requested):
1급 컬랙션으로 래핑하고, 도메인도 최대한 재사용과 조합을 사용하고자 하는 시도가 인상적이네요 👍

다만, 상속 구조에 대해서 문제가 발생할 소지가 있는 부분이 보입니다.
Card 라는 추상 클래스를 생성하고 `AlphabetCard`와 `NumberCard`를 각각 다른 구현체로 가지고갔는데요, 이 두 클래스가 서로 다른 성질을 가졌다고 판단하신것 같아요.

같은 부모클래스를 확장했다고 하더라도 클래스를 분리할 때에는 명확하게 '역할과 책임이 다른'것을 클래스로 분리해야 합니다. 영문카드와 숫자카드가 과연 역할과 책임이 다를까요?

예를 들자면, 카드중 A는 1 또는 11을 의미하며, J, Q, K는 각각 12, 13, 14로 치환이 가능합니다. 이 문자는 단순히 숫자를 표현하거나 출력하는 행위가 다른것이지, 상태(숫자)가 다른것은 아닙니다.

행위에 차이가 있는것은 상속이 아닌 인터페이스로 풀 수 있어요. 예를 들자면, `Printable` 이라는 인터페이스를 생성하고 이 인터페이스에 `print()`를 구현할 수 있도록 설정해 둔다면, 이 카드를 출력하는 부분에서 이 메서드를 통해 숫자카드든 영문카드든 동일하게 사용할 수 있지 않을까요?

f-lab-maverick commented:
참고로 제가 예시로 든 부분을 반영하여 구현하실 필요는 없습니다~ 참고만 해주세요

ZhenxiKim commented:
일급객체라는 개념에 대해 덕분에 알게되었습니다! 카드 상태가 변경되면 안돼서 필드를 final로 선언할까 하다가 그러면 초기화를 해줘야해서 어떻게 해야하나 했는데 해결방법으로 일급 객체를 사용에 대한 힌트 얻어갑니다!

Commit:
feat : 도메인 생성 변경

1. Card 상속 관계 제거
2. CardList 일급객체 제거
3. CardDeck 자료구조 스택으로 변경

gutenLEE reviewed (commented):
1. Card 상속 관계 제거
- Card 상속 관계를 다시 고민해보니 자식 클래스의 특별한 책임이나 역할이 없어 AlphabetCard, NumberCard 제거했습니다.
- Card 필드 CardType enum 추가했습니다.
- 알파벳인지 숫자 카드인지는 CardType enum 클래스에 카드 종류를 열거하고, 카드 번호로 CardType 을 결정하는 메서드 추가했습니다.

2. CardList 일급객체 제거
- CardDeck 이 현실 세계에서 카드 섞어주는 기계인줄 알고 기계와 카드묶음 클래스 두개를 설계했습니다.
다시 생각해보니 기계가 필요 없어서 CardList 클래스 제거했습니다.
- CardDeck 이 카드 리스트의 일급객체입니다.


3. CardDeck 자료구조 스택으로 변경
- 나눠준 카드는 제거해야하므로 자료구조를 List에서 Stack으로 변경했습니다.
- User 의 카드 List 를 Cards 일급객체 참조로 변경

f-lab-maverick commented on src/card/Card.java:
```diff
@@ -0,0 +1,29 @@
+ * @since 2023/06/18
+ */
+public class Card {
+    private final CardType cardType;
+    private final int number;
+    private final Suit suit;
```
final 키워드 사용으로 불변성을 보장해준 부분 아주 좋습니다.
생성자를 이용한 초기화를 해주셨네요 👍

f-lab-maverick commented on src/card/Card.java:
```diff
@@ -0,0 +1,29 @@
+
+        CardType type = CardType.getCardType(number);
+        if (cardType == UNDEFINED)
+            throw new RuntimeException("카드 번호 범위 초과");
```
예외처리는 명확한 예외 클래스를 만들어주세요. 모든 예외를 RuntimeException으로 처리한다면 예외처리문을 만들 때 어려움이 있습니다.

예를 들어, 정말 심각한 예외인지, 유효성만 실패해서 재시도를 하면 되는 예외인지도 `catch`문을 통해 구분하기 어려워집니다.

f-lab-maverick replied:
if문 내부 로직이 단 한줄이라도 중괄호를 이용하여 명확하게 코드블럭을 분리해주세요. 분리하지 않으면 if문 영향이 어디까지 가는지 헷갈릴 가능성이 생깁니다.

gutenLEE replied:
아하 반영하겠습니다

f-lab-maverick commented on src/card/CardType.java:
```diff
@@ -0,0 +1,34 @@
+
+    public static CardType getCardType(int number) {
+
+        if (number < 2 || number > 14)
```
약간의 팁을 드리자면 부등호 조건을 설정할 떄에는 읽기 편하도록 아래처럼 설정해주시는게 좋습니다. 사람의 뇌는 직관적으로 읽히는 방향이 아래와 같거든요.

1. `<`일 때에는 변수가 오른쪽
2. `>`일 때에는 변수가 왼쪽

```
if (number < 2 || 14 < number)
```

f-lab-maverick commented on src/card/Cards.java:
```diff
@@ -0,0 +1,16 @@
+public class Cards {
+    private final List<Card> cardList = new ArrayList<>();
+
+    public static Cards init() {
+        return new Cards();
+    }
```
팩토리 메서드를 만들어주셨네요. 생성자가 아닌 팩토리메서드를 init이라고 지어주신 이유가 있나요?
생성자도 그대로 남아있어서 동일한 역할이 아닌가 싶습니다. 어떤 의도로 만드신건가요?

gutenLEE replied:
Cards 에서 래핑하고 있는 cardList 를  필드초기화 하고 있기 때문에 
디폴트 생성자보다는 팩토리 메서드로  
"현재 보유하고 있는 카드가 없는 Cards 객체를 생성한다"는 의미를 전달하고 싶었습니다. 
`init()` 이라는 네이밍이 부적합했을까요?!

생성자는 삭제하겠습니다!

f-lab-maverick replied:
Factory Method 패턴을 사용할 때를 조금 정리해보자면
1. 생성시의 로직에 이름이 필요하거나
2. 생성자 오버로딩이 많거나
3. 미리 생성해놓는 캐싱 기법을 사용하거나
등등 여러 상황에서 사용합니다.

다만, 팩토리 메서드 패턴을 사용하지 않는 경우는 new 연산자와 정확하게 동일한 역할을 할 때 팩토리 메서드를 잘 사용하지 않습니다. 왜냐하면 팩토리 메서드를 사용함으로 얻을 수 있는 장점이 명확하게 없기 때문이에요.

이펙티브 자바 1장에서 다룰만큼 중요한 내용이기도 합니다. 아직 이펙티브자바는 읽지 않았으니 블로그에 정리해둔 내용이라도 한번 살펴보시겠어요? 현재처럼 팩토리메서드를 사용하며 이후 리팩토링해도 좋습니다.

f-lab-maverick commented on test/card/CardTypeTest.java:
```diff
@@ -0,0 +1,23 @@
+
+class CardTypeTest {
+
+    @Test
+    void getCardTypeTest() throws Exception {
+        Assertions.assertEquals(CardType.UNDEFINED, CardType.getCardType(1));
+        Assertions.assertEquals(CardType.NUMBER, CardType.getCardType(2));
+        Assertions.assertEquals(CardType.NUMBER, CardType.getCardType(5));
+        Assertions.assertEquals(CardType.NUMBER, CardType.getCardType(10));
+        Assertions.assertEquals(CardType.JACK, CardType.getCardType(11));
+        Assertions.assertEquals(CardType.QUEEN, CardType.getCardType(12));
+        Assertions.assertEquals(CardType.KING, CardType.getCardType(13));
+        Assertions.assertEquals(CardType.ACE, CardType.getCardType(14));
+        Assertions.assertEquals(CardType.UNDEFINED, CardType.getCardType(15));
+    }
```
첫 테스트 작성 아주 좋습니다!
테스트코드를 좀 더 작성해보시는걸 추천드리고, 테스트를 작성하시기 전 JUnit에 어떤 기능이 있는지 이 문서를 읽어보고 작성하시는걸 권장드립니다.
https://junit.org/junit5/docs/current/user-guide/

Commit:
feat : 도메인 생성 코드

Card.java
- 매개변수의 유효 범위 초과이므로 IllegalArgumentException 으로 변경

Cards.java
- private 생성자 변경
- 생성자 팩토리 메서드 이름 변경

Hand.java
- 클래스 이름 Cards 에서 Hand 로 변경
- 내가 가진 카드 패를 뜻하는 용어가 Hand 라고 합니다. 생성자 팩토리 메서드 이름을 고민하다 클래스명을 분명히 하고
팩토리 메서드 이름은 create 으로 일관성있게 하려했습니다

Commit:
feat : 도메인 생성 코드

Card.java
- 매개변수의 유효 범위 초과이므로 IllegalArgumentException 으로 변경

Cards.java
- private 생성자 변경
- 생성자 팩토리 메서드 이름 변경

Hand.java
- 클래스 이름 Cards 에서 Hand 로 변경
- 내가 가진 카드 패를 뜻하는 용어가 Hand 라고 합니다. 생성자 팩토리 메서드 이름을 고민하다 클래스명을 분명히 하고
팩토리 메서드 이름은 create 으로 일관성있게 하려했습니다

f-lab-maverick commented on src/card/Hand.java:
```diff
@@ -7,10 +7,12 @@
  * @since 2023/06/19
  */
-public class Cards {
+public class Hand {
```
블랙잭을 하는 현실세계를 표현하고자 `Hand`라는 클래스명을 사용한걸로 보입니다.
다만, 객체지향적인 코드를 작성할 때에는 현실과는 다르게 코드를 작성해야 더 좋은 코드가 나오는 경우가 많습니다.

이 클래스의 역할은 'User가 가진 카드를 모아놓는 클래스'입니다. Hand는 그 역할을 클래스명으로 표현하기는 어렵죠.

Spring에서는 이렇게 뭔가를 모아놓는 클래스를 만들 때에는 `ContextHolder`, `Provider` 등의 네이밍 규칙을 사용하고 있습니다. 이렇게 이름을 지으면 추후 이 클래스이름만 봐도 어떤 역할을 하는지 쉽게 유추할 수 있습니다.

이름을 짓는것에 대해 좀더 명시적으로 만들 수 있을지도 고민해보는게 좋을거같아요~

f-lab-maverick reviewed (approved):
고생하셨습니다. 머지하시고 다음 레벨의 요구사항을 구현해보세요~
